jsc.baseUrl causes canonicalization of symlinked source files, breaking bundler symlinks: false

### Describe the bug

When `jsc.baseUrl` is set in `builtin:swc-loader`, SWC unconditionally calls `canonicalize()` on file paths, resolving symlinks to their real filesystem paths. This defeats rspack's `resolve.symlinks: false` setting — relative imports from a symlinked source file resolve from the **real** path instead of the **symlink** location.

This affects any project that uses symlinked source files with `baseUrl` (e.g. monorepos sharing code via symlinks). Without `baseUrl`, the same setup works correctly because SWC does not create its `NodeImportResolver` and no canonicalization occurs.

Discovered via Meteor framework: https://github.com/meteor/meteor/issues/14053

### Input code

Minimal reproduction: **https://github.com/perbergland/swc-symlinks**

```
shared/
  source.ts          # imports ./dep (dep.ts does NOT exist here)
server/
  source.ts          # symlink -> ../shared/source.ts
  dep.ts             # exists here
src/
  index.ts           # imports ../server/source
```

`shared/source.ts`:
```typescript
import { VALUE } from "./dep";
export const myMethod = () => VALUE;
```

`server/dep.ts`:
```typescript
export const VALUE = "hello from server/dep";
```

With `resolve.symlinks: false`, the import `./dep` in `server/source.ts` (symlink) should resolve to `server/dep.ts`. Instead, SWC canonicalizes the path and looks for `shared/dep.ts`.

### Config

**Works** (no `baseUrl`):
```json
{
  "jsc": {
    "parser": { "syntax": "typescript" }
  }
}
```

**Fails** (with `baseUrl`):
```json
{
  "jsc": {
    "baseUrl": ".",
    "parser": { "syntax": "typescript" }
  }
}
```

### Playground link (or link to the minimal reproduction)

https://github.com/perbergland/swc-symlinks

```bash
git clone https://github.com/perbergland/swc-symlinks
cd swc-symlinks
npm install
npm run setup       # creates the symlink
npm run build       # works (no baseUrl)
npm run build:baseurl  # FAILS (with baseUrl)
```

### Expected behavior

Both builds succeed. `resolve.symlinks: false` should preserve symlink paths regardless of whether `jsc.baseUrl` is set.

### Actual behavior

`npm run build:baseurl` fails:

```
ERROR in ./shared/source.ts 1:0-30
  × Module not found: Can't resolve './dep' in '.../swc-symlinks/shared'
```

SWC resolves the symlink `server/source.ts` to its real path `shared/source.ts`, then `./dep` is looked up in `shared/` where it doesn't exist.

### Version

Reproduced with `@rspack/core` 1.6.7 (uses SWC internally via `builtin:swc-loader`).

### Additional context

**Root cause** — in [`swc_ecma_transforms_module/src/path.rs`](https://github.com/swc-project/swc/blob/main/crates/swc_ecma_transforms_module/src/path.rs) (around line 270):

```rust
// Bazel uses symlink
// https://github.com/swc-project/swc/issues/8265
if let FileName::Real(resolved) = &target.filename {
    if let Ok(orig) = canonicalize(resolved) {
        target.filename = FileName::Real(orig);
    }
}
```

`canonicalize()` unconditionally resolves symlinks. This was added to fix Bazel sandbox issues (#8265), but it breaks the common case of symlinked source files with `resolve.symlinks: false`.

**Why it only happens with `baseUrl`**: Without `baseUrl`, SWC skips creating its `NodeImportResolver` entirely, so no `canonicalize()` is called and symlink paths pass through unchanged.

**Related issues:**
- #8265 — Bazel sandbox symlinks (same `canonicalize()` code, different scenario)
- #4057 — Symlinked node_modules resolved incorrectly
- PR #6716 — Earlier fix attempt, reverted because it broke Bazel

**Suggested fix**: Add a `preserveSymlinks` option to the resolver config. When enabled, use `std::fs::read_link()` instead of `canonicalize()` — or skip canonicalization entirely — so that symlink paths are preserved. This would allow bundlers that set `resolve.symlinks: false` to work correctly while keeping the Bazel workaround available as the default behavior.

--- a/.changeset/good-lemons-study.md
+++ b/.changeset/good-lemons-study.md
@@ -1,0 +1,5 @@
+---
+swc_ecma_transforms_module: major
+---
+
+fix(es/module): add opt-in symlink-preserving resolver

--- a/crates/swc_ecma_transforms_module/Cargo.toml
+++ b/crates/swc_ecma_transforms_module/Cargo.toml
@@ -46,6 +46,7 @@ swc_ecma_visit = { version = "23.0.0", path = "../swc_ecma_visit" }
 [dev-dependencies]
 indexmap   = { workspace = true, features = ["serde"] }
 serde_json = { workspace = true }
+tempfile   = { workspace = true }
 
 swc_ecma_loader = { version = "21.0.0", path = "../swc_ecma_loader", features = [
   "node",

--- a/crates/swc_ecma_transforms_module/src/path.rs
+++ b/crates/swc_ecma_transforms_module/src/path.rs
@@ -94,6 +94,7 @@ where
 {
     resolver: R,
     config: Config,
+    preserve_symlinks: bool,
 }
 
 #[derive(Debug, Clone)]
@@ -118,6 +119,16 @@ where
     R: Resolve,
 {
     pub fn with_config(resolver: R, config: Config) -> Self {
+        Self::with_config_inner(resolver, config, false)
+    }
+
+    /// Creates a resolver that preserves symlink paths when rewriting module
+    /// specifiers.
+    pub fn with_config_preserving_symlinks(resolver: R, config: Config) -> Self {
+        Self::with_config_inner(resolver, config, true)
+    }
+
+    fn with_config_inner(resolver: R, config: Config, preserve_symlinks: bool) -> Self {
         #[cfg(not(all(target_arch = "wasm32", not(target_os = "wasi"))))]
         if let Some(base_dir) = &config.base_dir {
             assert!(
@@ -134,7 +145,11 @@ where
             );
         }
 
-        Self { resolver, config }
+        Self {
+            resolver,
+            config,
+            preserve_symlinks,
+        }
     }
 }
 
@@ -260,12 +275,14 @@ where
             }
         };
 
-        // Bazel uses symlink
-        //
-        // https://github.com/swc-project/swc/issues/8265
-        if let FileName::Real(resolved) = &target.filename {
-            if let Ok(orig) = canonicalize(resolved) {
-                target.filename = FileName::Real(orig);
+        if !self.preserve_symlinks {
+            // Bazel uses symlink
+            //
+            // https://github.com/swc-project/swc/issues/8265
+            if let FileName::Real(resolved) = &target.filename {
+                if let Ok(orig) = canonicalize(resolved) {
+                    target.filename = FileName::Real(orig);
+                }
             }
         }
 

--- a/crates/swc_ecma_transforms_module/tests/path_node.rs
+++ b/crates/swc_ecma_transforms_module/tests/path_node.rs
@@ -1,4 +1,5 @@
 use std::{
+    fs::{create_dir_all, write},
     path::{Path, PathBuf},
     sync::Arc,
 };
@@ -13,6 +14,7 @@ use swc_ecma_transforms_module::{
     rewriter::import_rewriter,
 };
 use swc_ecma_transforms_testing::{test_fixture, FixtureTestConfig};
+use tempfile::tempdir;
 use testing::run_test2;
 
 type TestProvider = NodeImportResolver<NodeModulesResolver>;
@@ -95,24 +97,83 @@ fn paths_resolver(
     rules: Vec<(String, Vec<String>)>,
     resolve_fully: bool,
 ) -> JscPathsProvider {
+    paths_resolver_with_options(base_dir, rules, resolve_fully, false)
+}
+
+fn paths_resolver_with_options(
+    base_dir: &Path,
+    rules: Vec<(String, Vec<String>)>,
+    resolve_fully: bool,
+    preserve_symlinks: bool,
+) -> JscPathsProvider {
     let base_dir = base_dir
         .to_path_buf()
         .canonicalize()
         .expect("failed to canonicalize");
     dbg!(&base_dir);
 
-    NodeImportResolver::with_config(
-        TsConfigResolver::new(
-            NodeModulesResolver::new(swc_ecma_loader::TargetEnv::Node, Default::default(), true),
-            base_dir.clone(),
-            rules,
-        ),
-        swc_ecma_transforms_module::path::Config {
-            base_dir: Some(base_dir),
-            resolve_fully,
-            file_extension: swc_ecma_transforms_module::util::Config::default_js_ext(),
-        },
-    )
+    let resolver = TsConfigResolver::new(
+        NodeModulesResolver::new(swc_ecma_loader::TargetEnv::Node, Default::default(), true),
+        base_dir.clone(),
+        rules,
+    );
+    let config = swc_ecma_transforms_module::path::Config {
+        base_dir: Some(base_dir),
+        resolve_fully,
+        file_extension: swc_ecma_transforms_module::util::Config::default_js_ext(),
+    };
+
+    if preserve_symlinks {
+        NodeImportResolver::with_config_preserving_symlinks(resolver, config)
+    } else {
+        NodeImportResolver::with_config(resolver, config)
+    }
+}
+
+#[test]
+fn symlink_paths_are_preserved_only_when_opted_in() {
+    let sandbox = tempdir().unwrap();
+    let root = sandbox.path();
+
+    create_dir_all(root.join("src")).unwrap();
+    create_dir_all(root.join("shared")).unwrap();
+    create_dir_all(root.join("server")).unwrap();
+
+    write(root.join("src/index.ts"), "import \"../server/source\";\n").unwrap();
+    write(root.join("shared/source.ts"), "export const value = 1;\n").unwrap();
+    create_symlink(
+        &root.join("shared/source.ts"),
+        &root.join("server/source.ts"),
+    );
+
+    let base = FileName::Real(root.join("src/index.ts").canonicalize().unwrap());
+
+    // Keep the original extensionless specifier so the assertion focuses on
+    // symlink preservation instead of extension rewriting.
+    let default_resolver = paths_resolver_with_options(root, Vec::new(), false, false);
+    let preserve_symlinks_resolver = paths_resolver_with_options(root, Vec::new(), false, true);
+
+    let default_resolved = default_resolver
+        .resolve_import(&base, "../server/source")
+        .unwrap();
+    let preserve_symlinks_resolved = preserve_symlinks_resolver
+        .resolve_import(&base, "../server/source")
+        .unwrap();
+
+    assert_eq!(&*default_resolved, "../shared/source");
+    assert_eq!(&*preserve_symlinks_resolved, "../server/source");
+}
+
+fn create_symlink(a: &Path, b: &Path) {
+    #[cfg(unix)]
+    {
+        std::os::unix::fs::symlink(a, b).unwrap();
+    }
+
+    #[cfg(windows)]
+    {
+        std::os::windows::fs::symlink_file(a, b).unwrap();
+    }
 }
 
 #[derive(Deserialize)]

